### PR TITLE
Refactoring useHttp to httpService

### PR DESCRIPTION
**Pull request by yy-wow**
English | [简体中文](https://github.com/opentiny/tiny-engine/blob/develop/.github/PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.zh-CN.md)

# PR

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our [Commit Message Guidelines](https://github.com/opentiny/tiny-engine/blob/develop/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Built its own designer, fully self-validated

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

### What is the current behavior?

Issue Number: N/A

### What is the new behavior?
支持用户自定义http请求配置
1. 增加MetaService：HttpService
2. 将 useHttp() 改为 getMetaApi(HttpService)
## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Release Notes

- **New Features**
	- Introduced a login popup for Single Sign-On (SSO) functionality.
	- Enhanced data fetching and error handling across various components using a new HTTP service.
	- Added `HttpService` to the engine's configuration for improved API interactions.

- **Improvements**
	- Updated methods for retrieving application IDs to streamline data access.
	- Enhanced user interaction with confirmation modals for critical actions like data source changes and deletions.
	- Improved validation logic for form submissions and error handling in asynchronous operations.

- **Dependency Updates**
	- Removed dependency on `@opentiny/tiny-engine-http` and added `axios` and `axios-mock-adapter` to improve HTTP handling.

- **Bug Fixes**
	- Resolved issues related to error handling and notifications during API calls.

### CODE REVIEW

File: designer-demo/src/composable/http/index.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+const preResponse = (res) => {
+  if (res.data?.error) {
+    showError(res.config?.url, res?.data?.error?.message)
+
+    return Promise.reject(res.data.error)
+  }
+
+  return res.data?.data
```
Comment (chilingling) [inline designer-demo/src/composable/http/index.js:49]: 可以看看  `data?.data` 能不能做成默认的一个拦截器

File: packages/common/composable/index.js [hunk 2]
```diff
@@ -1,2 +1,3 @@
 export { GenerateCodeService } from './generateCode'
 export { default as GlobalService } from './defaultGlobalService'
+export { default as HttpService } from './http'
```
[the remark this answers is not in the document]
Comment (yy-wow) [reply packages/common/composable/index.js:3]: 为 ./http/index.js 省略

File: packages/toolbars/logout/src/js/http.js [hunk 3]
```diff
@@ -10,9 +10,7 @@
...
  */
 
-import { useHttp } from '@opentiny/tiny-engine-http'
-
-const http = useHttp()
+import { getMetaApi, META_SERVICE } from '@opentiny/tiny-engine-meta-register'
 
 // 注销
-export const requestLogout = () => http.post(`platform-center/api/logout`)
+export const requestLogout = () => getMetaApi(META_SERVICE.Http).post(`platform-center/api/logout`)
```
[the remark this answers is not in the document]
Comment (yy-wow) [reply packages/toolbars/logout/src/js/http.js:16]: Errors are handled by the user

File: packages/plugins/state/src/js/http.js [hunk 4]
```diff
@@ -10,11 +10,10 @@
...
-const http = useHttp()
+import { getMetaApi, META_SERVICE } from '@opentiny/tiny-engine-meta-register'
 
 // 状态管理 —— 更新全局变量
-export const updateGlobalState = (appId, params) => http.post(`/app-center/api/apps/update/${appId}`, params)
+export const updateGlobalState = (appId, params) =>
+  getMetaApi(META_SERVICE.Http).post(`/app-center/api/apps/update/${appId}`, params)
 
-export const requestGlobalState = (appId) => http.get(`/app-center/api/apps/schema/${appId}`)
+export const requestGlobalState = (appId) => getMetaApi(META_SERVICE.Http).get(`/app-center/api/apps/schema/${appId}`)
```
[the remark this answers is not in the document]
Comment (yy-wow) [reply packages/plugins/state/src/js/http.js:19]: Errors are handled by the user

File: packages/toolbars/generate-code/src/http.js [hunk 5]
```diff
@@ -29,13 +27,14 @@ export const fetchCode = async ({ platform, app, pageInfo, tenant } = {}) =>
...
-    ? http.get('/app-center/api/preview/metadata', {
+    ? getMetaApi(META_SERVICE.Http).get('/app-center/api/preview/metadata', {
         headers: { [HEADER_LOWCODE_ORG]: tenant },
         params: { platform, app, type, id, history }
       })
     : {}
 
 // 获取页面列表
-export const fetchPageList = (appId) => http.get(`/app-center/api/pages/list/${appId}`)
+export const fetchPageList = (appId) => getMetaApi(META_SERVICE.Http).get(`/app-center/api/pages/list/${appId}`)
```
[the remark this answers is not in the document]
Comment (yy-wow) [reply packages/toolbars/generate-code/src/http.js:37]: Errors are handled by the user

File: packages/toolbars/generate-code/src/http.js [hunk 6]
```diff
@@ -29,13 +27,14 @@
...
 
-export const fetchBlockSchema = async (blockName) => http.get(`/material-center/api/block?label=${blockName}`)
+export const fetchBlockSchema = async (blockName) =>
+  getMetaApi(META_SERVICE.Http).get(`/material-center/api/block?label=${blockName}`)
```
[the remark this answers is not in the document]
Comment (yy-wow) [reply packages/toolbars/generate-code/src/http.js:40]: Errors are handled by the user

File: designer-demo/src/composable/http/mock.js [hunk 7]
```diff
@@ -0,0 +1,939 @@
...
+    url: /\/app-center\/api\/apps\/list/,
+    response: async () => {
+      const getAppList = async () => {
+        const response = await fetch('/mock/appList.json')
+        return response.json()
+      }
+
+      const appListRes = await getAppList()
+      return new Promise((resolve) => {
+        setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (yy-wow) [reply designer-demo/src/composable/http/mock.js]: handled by user

File: designer-demo/src/composable/http/index.js (hunk 1 above)
Comment (yy-wow) [reply designer-demo/src/composable/http/index.js:49]: response包含多个信息，暴露给用户自行处理
image

File: designer-demo/src/composable/http/mock.js [hunk 8]
```diff
@@ -0,0 +1,939 @@
...
+  {
+    url: /\/app-center\/api\/apps\/detail/,
+    response: async () => {
+      const getAppInfo = async () => {
+        const response = await fetch('/mock/appInfo.json')
+        return response.json()
+      }
+
+      const appInfoRes = await getAppInfo()
+      return new Promise((resolve) => {
```
[the remark this answers is not in the document]
Comment (yy-wow) [reply designer-demo/src/composable/http/mock.js:30]: handled by user

File: designer-demo/src/composable/http/index.js (hunk 1 above)
Comment (yy-wow) [reply designer-demo/src/composable/http/index.js:49]: 在cli的template中添加默认配置拦截器

File: packages/common/composable/defaultGlobalService.js [hunk 9]
```diff
@@ -43,18 +42,18 @@ const initialState = {
...
     })
 }
 
 // 获取当前应用的信息
-const fetchAppInfo = (appId) => useHttp().get(`/app-center/api/apps/detail/${appId}`)
+const fetchAppInfo = (appId) => getMetaApi(META_SERVICE.Http).get(`/app-center/api/apps/detail/${appId}`)
 
 // 获取应用列表
-const fetchAppList = (platformId) => useHttp().get(`/app-center/api/apps/list/${platformId}`)
+const fetchAppList = (platformId) => getMetaApi(META_SERVICE.Http).get(`/app-center/api/apps/list/${platformId}`)
```
[the remark this answers is not in the document]
Comment (yy-wow) [reply packages/common/composable/defaultGlobalService.js:56]: 错误处理交给用户自行处理

File: designer-demo/src/composable/http/index.js
Comment (chilingling) [inline designer-demo/src/composable/http/index.js:1]: 这块 `composable `  文件夹的命名可以考虑下

File: packages/common/composable/http/axios.js [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+ * Use of this source code is governed by an MIT-style license.
+ *
+ * THE OPEN SOURCE SOFTWARE IN THIS PRODUCT IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
+ * BUT WITHOUT ANY WARRANTY, WITHOUT EVEN THE IMPLIED WARRANTY OF MERCHANTABILITY OR FITNESS FOR
+ * A PARTICULAR PURPOSE. SEE THE APPLICABLE LICENSES FOR MORE DETAILS.
+ *
+ */
+
+import axios from 'axios'
+import MockAdapter from 'axios-mock-adapter'
```
Comment (chilingling) [inline packages/common/composable/http/axios.js]: 其实 mock 的功能也可以考虑去掉，用户的工程需要用到用户自己实现（其实现在大部分用户也不会用到这个 mock 的工程）。干掉 Mock 的功能，这个 axios 的文件应该都可以干掉了，只需要 `import axios from 'axios'`

File: packages/common/composable/http/index.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    const http = axios(axiosConfig)
+
+    enableMock && http.mock(mockData)
+
+    const { request = [], response = [] } = interceptors
+
+    addInterceptors({ data: request, http, type: 'request' })
+    addInterceptors({ data: response, http, type: 'response' })
+
+    state.http = http
```
Comment (chilingling) [inline packages/common/composable/http/index.js]: 这里的 `http` 其实不是一个响应式的状态变量，是不是用闭包存储起来就可以了？没必要变成一个 `reactive` 对象？

File: packages/common/composable/http/axios.js (hunk 10 above)
Comment (yy-wow) [reply packages/common/composable/http/axios.js]: 已删除

File: packages/common/composable/http/index.js (hunk 11 above)
Comment (yy-wow) [reply packages/common/composable/http/index.js]: 已修改

File: designer-demo/src/composable/http/index.js
Comment (yy-wow) [reply designer-demo/src/composable/http/index.js:1]: 暂不修改

**Review by chilingling — DISMISSED** [review]

File: designer-demo/src/composable/http/index.js [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+const { post: globalNotify } = useBroadcastChannel({ name: BROADCAST_CHANNEL.Notify })
+
+const procession = {
+  promiseLogin: null,
+  mePromise: {}
+}
+
+const loginDom = document.createElement('div')
+document.body.appendChild(loginDom)
```
Comment (rhlin) [inline designer-demo/src/composable/http/index.js]: 顶层最好呢是不要有自执行的代码。

遵循模块编程的思想， 除了入口模块（如main.js），其他模块不要有顶层自执行代码，提升代码模块的兼容性，去掉模块引用立即产生的副作用。

File: designer-demo/src/composable/http/index.js [hunk 13]
```diff
@@ -67,6 +44,29 @@ const preResponse = (res) => {
 }
 
 const openLogin = () => {
+  const procession = {
+    promiseLogin: null,
+    mePromise: {}
+  }
+
+  const loginDom = document.createElement('div')
```
Comment (chilingling) [inline designer-demo/src/composable/http/index.js]: 这里做个单例？不需要每次都创建？

File: designer-demo/src/composable/http/index.js (hunk 12 above)
Comment (yy-wow) [reply designer-demo/src/composable/http/index.js]: 已修改

File: designer-demo/src/composable/http/index.js (hunk 13 above)
Comment (yy-wow) [reply designer-demo/src/composable/http/index.js]: 已修改

**Review by chilingling — DISMISSED** [review]

File: designer-demo/src/composable/http/mock.js [hunk 14]
```diff
@@ -702,10 +691,11 @@ export default [
       const blocks = await getAvailable()
       const tags = []
 
-      blocks &&
-        blocks.forEach((block) => {
-          block.tags && tags.push(...block.tags)
-        })
+      blocks?.forEach((block) => {
+        if (block.tags?.length) {
```
Comment (gene9831) [inline designer-demo/src/composable/http/mock.js]: 当 `block.tags` 数组长度为0时，不会进入 `if` 内部逻辑。建议更改为 `Array.isArray` 来进行判断，其他地方类似代码也是如此

File: designer-demo/src/composable/http/index.js [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+const customizeHttpService = () => {
+  const options = {
+    axiosConfig: getConfig(),
+    interceptors: {
+      request: [preRequest],
+      response: [[preResponse, errorResponse]]
+    }
+  }
+
+  HttpService.apis.setOptions(options)
```
Comment (gene9831) [inline designer-demo/src/composable/http/index.js:130]: 这里应该可以直接修改 `HttpService.options` 来传入自定义的options，是否有必要增加一个 `setOptions` api

Comment (yy-wow) [reply designer-demo/src/composable/http/index.js:130]: setOptions是将传入的配置与默认配置合并，适用于多项配置一次性修改，如果只是少量配置修改，可以直接HttpService.options.xxx来覆盖

File: designer-demo/src/composable/http/mock.js (hunk 14 above)
Comment (yy-wow) [reply designer-demo/src/composable/http/mock.js]: mock文件删除，交给用户自行决定是否mock

**Review by hexqi — DISMISSED** [review]

**Review by hexqi — COMMENTED** [review]
出码那块没有修改，还是之前的http文件

File: packages/common/package.json [hunk 16]
```diff
@@ -43,12 +42,14 @@
...
     "monaco-editor": "0.33.0",
     "prettier": "2.7.1",
+    "axios": "~0.28.0",
     "vue-draggable-next": "2.1.0"
   },
   "devDependencies": {
     "@opentiny/tiny-engine-vite-plugin-meta-comments": "workspace:*",
     "@vitejs/plugin-vue": "^5.1.2",
     "@vitejs/plugin-vue-jsx": "^4.0.1",
+    "axios-mock-adapter": "^1.21.5",
```
Comment (hexqi) [inline packages/common/package.json]: http去掉了mock能力，npm依赖也要去掉

Comment (yy-wow) [reply packages/common/package.json]: 已删除

**Review by chilingling — APPROVED** [review]

**Pull request merged**